### PR TITLE
[Modules] Multi additions to staticSites

### PR DESCRIPTION
**Pull request by itpropro**
# Description

Added the following to staticSites

- `config` (app config and function app config)
- `customDomains` with different validations depending on apex or not
- `linkedBackends` with the 2022-03-01 API to add custom function apps. The `userDefinedFunctions` endpoint should not be used and is currently broken.

Changes were verified, pester and deployment tested with the respective PowerShell functions. Readme and parameter files have been updated.

## Pipeline references

No pipeline changes

# Type of Change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] Update to documentation

# Checklist

- [X] I'm sure there are no other open Pull Requests for the same update/change
- [X] My corresponding pipelines / checks run clean and green without any errors or warnings
- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (readme)
- [X] I did format my code

### CODE REVIEW

File: modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep [hunk 1]
```diff
@@ -4,9 +4,9 @@ param roleDefinitionIdOrName string
 param resourceId string
 
 var builtInRoleNames = {
-  'Owner': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')
-  'Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b24988ac-6180-42a0-ab88-20f7382dd24c')
-  'Reader': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+  Owner: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep:7]: Why this change?

File: modules/Microsoft.Web/staticSites/config/readme.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+**Conditional parameters**
+| Parameter Name | Type | Description |
+| :-- | :-- | :-- |
+| `staticSiteName` | string | The name of the parent Static Web App. Required if the template is used in a standalone deployment. |
+
+
+### Parameter Usage: `<ParameterPlaceholder>`
+
+// TODO: Fill in Parameter usage
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/config/readme.md:34]: Please remove section unless useful to be customized

File: modules/Microsoft.Web/staticSites/config/readme.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# Web StaticSites Config `[Microsoft.Web/staticSites/config]`
+
+This module deploys Web StaticSites Config.
+// TODO: Replace Resource and fill in description
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/config/readme.md:4]: I'd suggest the following unless additional description is useful to be added.
```suggestion
This module deploys Web Static Sites Config.

```

File: modules/Microsoft.Web/staticSites/config/deploy.bicep [hunk 4]
```diff
@@ -0,0 +1,33 @@
+@allowed([
+  'appsettings'
+  'functionappsettings'
+])
+@description('Required. Type of settings to apply.')
+param kind string
+
+@description('Required. App settings.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/config/deploy.bicep:8]: Would it make sense to pass properties as separated parameters for better parameter validation?

File: modules/Microsoft.Web/staticSites/deploy.bicep [hunk 5]
```diff
@@ -65,7 +65,7 @@ param userAssignedIdentities object = {}
 @description('Optional. Specify the type of lock.')
 param lock string = ''
 
-@description('Optional. Configuration details for private endpoints. For security reasons, it is recommended to use private endpoints whenever possible.')
+@description('Optional. Configuration details for private endpoints.')
```
Comment (eriqua) [inline modules/Microsoft.Web/staticSites/deploy.bicep:68]: Please keep the removed sentence for consistency with other modules

File: modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep (hunk 1 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep:7]: Hey @eriqua, this is aligned with Bicep best practice and avoids this warning message:

```
Warning prefer-unquoted-property-names: Property names that are valid identifiers should be declared without quotation marks and accessed using dot notation. [https://aka.ms/bicep/linter/prefer-unquoted-property-names]
```

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 3 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/config/readme.md:4]: I would suggest to either use the API name (`StaticSites`) or the product name (`Static Web App`).

File: modules/Microsoft.Web/staticSites/deploy.bicep (hunk 5 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/deploy.bicep:68]: I always use the field description generated in the API documentation with autorest from the official specs. Please add any CARML specific deviations to our docs here :)

File: modules/Microsoft.Web/staticSites/config/deploy.bicep (hunk 4 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep:8]: I am not sure how it would improve parameter validation. The object passed for the settings themselves has to be a object in both cases, that currently cannot be validated by bicep anyway. Additionally for a deployment within the StaticSites resource it is abstracted into two separate deployments anyway, see 144, 153 in the main deploy of StaticSites.

File: modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep (hunk 1 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep:7]: Please undo, this is a new warning message with is disabled with PR #1632.

Comment (itpropro) [reply modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep:7]: Why are we not complying with Bicep best practices here @MrMCake? When just importing some of the modules in a project, the current `bicepconfig.json` is probably not going to be copied or has it's own bicepconfig.json. Also how does this behave with CARML modules consumed from a ACR?
I think we should comply with the Bicep linter wherever possible, I don't see any benefit in deviating, but downsides for all CARML customers that will receive warnings in their projects.

Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep:7]: The rule was just introduced, which means all modules now produce this warning. 
No matter whether one thinks the rule is good or not (in this case I, for example, prefer that all roles use the same syntax), I'd much rater prefer consistency across the modules and change them all at ones instead of one module at a time.

File: modules/Microsoft.Web/staticSites/deploy.bicep (hunk 5 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/deploy.bicep:68]: @eriqua + 1.

While we (most of the time) take the description from the API as well, we sometimes add further information like in this case. So the ask would be to just not remove the extra description :)

File: modules/Microsoft.Web/staticSites/config/deploy.bicep (hunk 4 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/deploy.bicep:8]: Agreed. It's not like there is any documentation anyways :)

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 3 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/readme.md:4]: Why was this comment resolved? I can still only see the placeholder.

File: modules/Microsoft.Web/staticSites/customDomains/readme.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+# Web StaticSites CustomDomains `[Microsoft.Web/staticSites/customDomains]`
+
+This module deploys Web StaticSites CustomDomains.
+// TODO: Replace Resource and fill in description
```
Comment (MrMCake) [inline modules/Microsoft.Web/staticSites/customDomains/readme.md:4]: Please replace the placeholder. For example with `This module deploys custom domains for static sites.`

File: modules/Microsoft.Web/staticSites/linkedBackends/readme.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# Web StaticSites LinkedBackends `[Microsoft.Web/staticSites/linkedBackends]`
```
Comment (MrMCake) [inline modules/Microsoft.Web/staticSites/linkedBackends/readme.md:1]:
```suggestion
# Static Site Linked Backend `[Microsoft.Web/staticSites/linkedBackends]`
```

File: modules/Microsoft.Web/staticSites/customDomains/readme.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
+# Web StaticSites CustomDomains `[Microsoft.Web/staticSites/customDomains]`
```
Comment (MrMCake) [inline modules/Microsoft.Web/staticSites/customDomains/readme.md:1]:
```suggestion
# Static Site Custom Domain `[Microsoft.Web/staticSites/customDomains]`

```

File: modules/Microsoft.Web/staticSites/config/readme.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
+# Web StaticSites Config `[Microsoft.Web/staticSites/config]`
```
Comment (MrMCake) [inline modules/Microsoft.Web/staticSites/config/readme.md:1]:
```suggestion
# Static Site Config `[Microsoft.Web/staticSites/config]`

```

File: modules/Microsoft.Web/staticSites/config/readme.md (hunk 3 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/config/readme.md:4]: My suggestion:
```suggestion
This module deploys a Static Site Config.
```

File: modules/Microsoft.Web/staticSites/deploy.bicep (hunk 5 above)
Comment (itpropro) [reply modules/Microsoft.Web/staticSites/deploy.bicep:68]: I just blindly copied the text from the specs, this has slipped :D

**Comment by itpropro** [comment]
PR feedback integrated, thx @eriqua and @MrMCake!

File: modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep (hunk 1 above)
Comment (MrMCake) [reply modules/Microsoft.Web/staticSites/.bicep/nested_roleAssignments.bicep]: I'd suggest a new issue as also #1632 does not involve the main source of the warnings: the RBAC roles ;)

**Review by MrMCake — APPROVED** [review]
Approved with merge to branch `users/itpropro/staticSites` for pipeline validation
